interception-tools: init at 0.1.1

vyp opened the pull request:
The latest release of libyamlcpp in nixpkgs does not build because it
uses an older version of boost than the one in nixpkgs and therefore
expects a particular header file which does not exist in the latest
boost anymore. For this reason, a later (git) version of libyamlcpp is
used here (which actually doesn't even require boost).

The substituteInPlace in the prePatch phase is needed because libevdev
places its headers in non-standard places, meaning Nix cannot normally
find them. The `cut` command removes the first two "-I" characters from
the output of `pkg-config`. This needs to be in the prePatch phase
because otherwise Nix will patch these lines to `/var/empty`, meaning
you would have less specific replacement (in case other lines are also
patched to `/var/empty`).

I wrote the patch. (I believe it is NixOS specific.)

###### Motivation for this change

I want to use this:

- https://gitlab.com/interception/linux/tools
- https://gitlab.com/interception/linux/plugins/caps2esc

I wasn't sure if I should have changed libyamlcpp in nixpkgs instead, because it doesn't have a new release (but this package requires a newer version). And also I'm not sure if updating it would break packages using it.. how are such library updates handled in nixpkgs? 😕

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Mic92 commented on nixos/modules/services/hardware/interception-tools.nix:
```diff
@@ -0,0 +1,60 @@
+{ config, lib, pkgs, ... }:
```
You also need to add this module to `nixos/modules/module-list.nix` as well.

vyp replied:
Thank you, fixed. I missed this because although I tested the packages themselves, I didn't test the service and just copied the file over, sorry. :expressionless:

vyp replied:
As in, I did test the service, just not from under nixpkgs, but from importing it in `configuration.nix`.

Mic92 commented on nixos/modules/services/hardware/interception-tools.nix:
```diff
@@ -0,0 +1,60 @@
+      '';
+      path = [ pkgs.bash pkgs.interception-tools ] ++ cfg.plugins;
+      script = ''
+        nice -n -20 udevmon -c \
```
The `Nice=` property of systemd can do the same. Then `script` can be replaced by `service.ExecStart`.

Mic92 replied:
Why is it actually necessary to change the nice level at all? Shouldn't user input be processed as fast as possible?

vyp replied:
Thanks, updated. I didn't know about `Nice`.

And yeah if you check the manpage for `systemd.exec` and search for nice, it says that -20 is the highest priority. But I just got this from the [package readme](https://gitlab.com/interception/linux/tools#execution), where it says that it increases the `udevmon` priority since it'll be responsible for a vital input device, just to make sure it stays "responsible".

Mic92 commented on pkgs/tools/inputmethods/interception-tools/default.nix:
```diff
@@ -0,0 +1,49 @@
+  libudev }:
+
+let
+  libyamlcppWithoutBoost = libyamlcpp.overrideAttrs (oldAttrs: rec {
```
This means, it also needs `libyaml` without boost and not just a newer version?

vyp replied:
I'm assuming you meant `libyamlcpp` because `libyaml` is for C, so it can't use boost anyway since boost is a c++ library.

But yeah so `libyamlcpp` is just a dependency for `interception-tools`, and `boost` is a dependency for the latest release of `libyamlcpp`. So `interception-tools` doesn't directly need boost at all, if that makes sense. However, the boost problem obviously disappears if I use a more recent *git* version of libyamlcpp instead, because they got rid of the boost dependency since their last release. So I thought that I should just do that. (And also, the latest release of libyamlcpp was in January 2016 too, so it's been a while.)

I guess, are you saying that `libyamlcppWithoutBoost` is a misleading name? In that case, maybe `libyamlcppGit` then? Or any suggestions? I don't know if there is a convention here.

Mic92 replied:
I thought, there is some kind of optional boost support in the library.

Mic92 replied:
I would move this expression next to `libyamlcpp` in `all-packages.nix`, with a comment that it is required by `interception-tools`. This way, we can easier remove it again, when there is a `libyamlcpp` update.

vyp replied:
Ok, done. :+1:

vyp committed:
interception-tools: init at 0.1.1

The latest release of libyamlcpp in nixpkgs does not build because it
uses an older version of boost than the one in nixpkgs and therefore
expects a particular header file which does not exist in the latest
boost anymore. For this reason, a later (git) version of libyamlcpp is
used here (which actually doesn't even require boost).

The substituteInPlace in the prePatch phase is needed because libevdev
places its headers in non-standard places, meaning Nix cannot normally
find them. The `cut` command removes the first two "-I" characters from
the output of `pkg-config`. This needs to be in the prePatch phase
because otherwise Nix will patch these lines to `/var/empty`, meaning
you would have less specific replacement (in case other lines are also
patched to `/var/empty`).

I wrote the patch. (I believe it is NixOS specific.)

Mic92 commented:
Thanks!
